### PR TITLE
o/configstate/configcore: support SSH socket activation configurations

### PR DESCRIPTION
**Pull request by Meulengracht**
On ubuntu 22.10+ sshd uses socket activation and that means we must reload SSH differently for the socket unit.

This should fix the nested test tests/nested/manual/core20-early-config

REF: SNAPDENG-34249

### CODE REVIEW

File: overlord/configstate/configcore/services_test.go
Comment (valentindavid) [inline overlord/configstate/configcore/services_test.go:493]: Should this one has `FreeBind=true`?

Comment (Meulengracht) [reply overlord/configstate/configcore/services_test.go:493]: No not really needed, it is only needed for IP addresses, but i added it in all cases. I'll get this properly designed

File: overlord/configstate/configcore/services.go
Comment (ernestl) [inline overlord/configstate/configcore/services.go:376]: Super nitpick, can this be part of previous line?

File: overlord/configstate/configcore/services.go
Comment (ernestl) [inline overlord/configstate/configcore/services.go:379]: Presume this is not applicable anymore: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1819728

**Review by ernestl — APPROVED** [review]
Thanks

File: overlord/configstate/configcore/services.go
Comment (Meulengracht) [reply overlord/configstate/configcore/services.go:379]: Nope, atleast not from observations, they seem to indicate things are fine

File: overlord/configstate/configcore/services.go
Comment (Meulengracht) [reply overlord/configstate/configcore/services.go:376]: I'll move it up when doing a final revision, will just wait for another review - thank you!

File: overlord/configstate/configcore/services.go
Comment (ernestl) [inline overlord/configstate/configcore/services.go:386]: Should this not be done always regardless of requiring daemon reload, to ensure we sync service config?

**Review by ernestl — COMMENTED** [review]
On second thought, just to double check

File: overlord/configstate/configcore/services.go
Comment (Meulengracht) [reply overlord/configstate/configcore/services.go:386]: Nope, the other approach is documented exactly to do it like that - and it works by restarting the socket. The reason it works is because of `daemon-reload`, because the socket is actually generated by a systemd generator

Comment (Meulengracht) [reply overlord/configstate/configcore/services.go:386]: This is from the documentation:
```
# When systemd socket activation is used (the default), the socket
# configuration must be re-generated after changing Port, AddressFamily, or
# ListenAddress.
#
# For changes to take effect, run:
#
# systemctl daemon-reload
# systemctl restart ssh.socket
```

**Review by zyga — APPROVED** [review]
Looks good. Could have this been caught by a non-nested test perhaps? I wonder why we found it so late given that this is 22.10+

**Comment by Meulengracht** [comment]
> Looks good. Could have this been caught by a non-nested test perhaps? I wonder why we found it so late given that this is 22.10+

For why found it so late: This functionality (the option) is core only, which means that it works fine on UC22, but the bug only appeared in UC24. This test has been failing since we introduced the test for UC24.

We could probably cover this in a non-nested way, I believe the reason why it's done like this is because of the SSH access changing. I.e if you reconfigure the SSH access in a non-nested way, then spread would lose connectivity to the VM. (Is my guess)

File: overlord/configstate/configcore/services.go
Comment (ernestl) [reply overlord/configstate/configcore/services.go:385]: image
image

Comment (Meulengracht) [reply overlord/configstate/configcore/services.go:385]: Yea I was wrong in regards to that, but following the documentation provided for making changes take effect works, and let us have that as-is.

**Review by ernestl — APPROVED** [review]

**Pull request merged**
